import-areas/import-areas.go

```diff
@@ -226,11 +239,25 @@ func postAreaQuery(query string) ([]models.Area, error) {
 	v.Set("query", query)
 
 	client := &http.Client{}
-	req, _ := http.NewRequest("POST", "http://statistics.data.gov.uk/sparql", strings.NewReader(v.Encode()))
+	req, err := http.NewRequest("POST", "http://statistics.data.gov.uk/sparql", strings.NewReader(v.Encode()))
+	if err != nil {
+		log.Event(ctx, "error returned from building an area request", log.ERROR, log.Error(err))
+		return nil, err
+	}
+
 	req.Header.Add("Accept", "application/sparql-results+json")
-	resp, _ := client.Do(req)
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Event(ctx, "error returned from the POST request", log.ERROR, log.Error(err))
+		return nil, err
+	}
```

Comment:
It's good practice to put your `defer resp.Body.Close()` statements immediately after the `err != nil check`. This will ensure that `close` is called - regardless of success or failure. 

I've checked the full file for this change and I can see you've added a defer close **but** it's after the  `ioutil.ReadAll(resp.Body)` - if `ReadAll` fails the defer will never be called and we could end up in a state where we're not closing a resource properly which we don't want.

My personal rule of thumb is to always add any `defer close()` /clean up immediately **before** doing the thing that needs to be cleaned up - it's very easy to overlook & this helps to remember to do it.

Reply:
Thank you for sharing this

Reply:
I know I am being pedant about this but the `defer resp.Body.Close()` needs to come immediately after checking the error and _before_ using the response object,

The [http.Client godoc](https://golang.org/pkg/net/http/#Client.Do) states:

> If the returned error is nil, the Response will contain a non-nil Body which the user is expected to close. If the Body is not both read to EOF and closed, the Client's underlying RoundTripper (typically Transport) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.

In this code:  if `status != 200` then function will return before a `defer resp.Body.Close()` is registered which means that the response body resource will not be read & closed which had the impact outline above.